File: LintCode-Java/SingleNumber.java
public class SingleNumber {
    /**
     *@param A : an integer array
     *return : a integer
     */

    // V1, O(n), O(1)
    // XOR
    public int singleNumber(int[] A) {
        if (A == null || A.length == 0) {
            return 0;
        }

        int n = A[0];
        for(int i = 1; i < A.length; i++) {
            n = n ^ A[i];
        }

        return n;
    }
}

File: LintCode-Python/single_number.py
"""
Given 2*n + 1 numbers, every numbers occurs twice except one, find it.

Example
Given [1,2,2,1,3,4,3], return 4

Challenge
One-pass, constant extra space.
"""


class Solution:
    """
    @param A : an integer array
    @return : a integer
    """
    def singleNumber(self, A):
        # O(n), O(1)
        if not A:
            return 0
        res = A[0]
        for i in range(1, len(A)):
            res ^= A[i]
        return res
